gridstatus/isone_api/isone_api.py

```diff
@@ -0,0 +1,379 @@
+                        f"Error: Rate-limited still after {self.max_retries} retries. "
+                        f"Failed to get data from {url} with params: {api_params}"
+                    )
+                else:
+                    error_message = (
+                        f"Error: Failed to get data from {url} with params:"
+                        f" {api_params}"
+                    )
+                log.error(error_message)
+                response.raise_for_status()
+
+        if parse_json:
+            return response.json()
+        else:
+            return response.content
+
+    def get_locations(self) -> pd.DataFrame:
+        """
+        Get a list of core hub and zone locations.
+
+        Returns:
+            pandas.DataFrame: A DataFrame containing location information.
+        """
+        url = f"{BASE_URL}/locations"
+        response = self.make_api_call(url)
+        if "Locations" not in response or "Location" not in response["Locations"]:
+            raise NoDataFoundException("No location data found.")
+
+        locations = response["Locations"]["Location"]
+        df = pd.DataFrame(locations)
+
+        return df
+
+    def get_location_by_id(self, location_id: int) -> pd.DataFrame:
+        """
+        Get information for a specific location by its ID.
+
+        Args:
+            location_id (int): The ID of the location to retrieve.
+
+        Returns:
+            pandas.DataFrame: A DataFrame containing the location information.
+        """
+        url = f"{BASE_URL}/locations/{location_id}"
+        response = self.make_api_call(url)
+
+        if "Location" not in response:
+            raise NoDataFoundException(f"No data found for location ID: {location_id}")
+
+        location = response["Location"]
+        df = pd.DataFrame([location])
+        return df
+
+    def get_locations_all(self) -> pd.DataFrame:
+        """
+        Get detailed information for all locations.
+
+        Returns:
+            pandas.DataFrame: A DataFrame containing detailed information for all locations.
+        """
+        url = f"{BASE_URL}/locations/all"
+        response = self.make_api_call(url)
+
+        if "Locations" not in response or "Location" not in response["Locations"]:
+            raise NoDataFoundException("No location data found.")
+
+        locations = response["Locations"]["Location"]
+        df = pd.DataFrame(locations)
+        return df
+
+    def _handle_demand(
+        self,
+        df: pd.DataFrame,
+        interval_minutes: int = 60,
+    ) -> pd.DataFrame:
+        """
+        Process demand DataFrame: convert types, rename columns, and add Interval End.
+
+        Args:
+            df (pd.DataFrame): Input DataFrame with demand data.
+            interval_minutes (int): Duration of each interval in minutes. Default is 60.
+
+        Returns:
+            pd.DataFrame: Processed DataFrame.
+        """
+        df["Interval Start"] = pd.to_datetime(df["BeginDate"])
```

Comment:
i think we want to do `pd.to_datetime(df["BeginDate"]).dt.tz_convert(self.default_timezone)` to make sure this gets but into the right timezone.

it's minor difference but it takes it from the absolute `-4` offset to the named `US/Eastern` that respects DST.

The situation where this matter is when the interval end is on the other side of a DST change, so having the fixed offset is confusion. 

<img width="646" alt="image" src="https://github.com/user-attachments/assets/a29ea5d3-9676-46e0-82e2-1de052c97876">